### PR TITLE
PHP 8.3 Support: Typed class constants #6701

### PR DESCRIPTION
**Pull request by junichi11**
- https://github.com/apache/netbeans/issues/6701
- https://wiki.php.net/rfc/typed_class_constants

### PHP 8.3 Support: Typed class constants (Part 1)

- Fix the grammar file (parser)
- Fix the `PHP83UnhandledError`
- Add/Fix unit tests for the parser
- Increment the spec version

Example:
```php
class Example {
    public const string CONSTANT1 = 'constant';
    public const ?int CONSTANT2 = 1;
    public const int|string CONSTANT3 = 1;
}
```

PHP 8.3

nb-php83-typed-class-constants-parser

PHP 8.2

nb-php83-typed-class-constants-parser-php82

### PHP 8.3 Support: Typed class constants (Part 2)

- Index class constant types
- Fix the navigator
- Add/Fix unit tests for the indexer and the navigator
- Increment the spec version

nb-php83-typed-class-constants-navigator

### PHP 8.3 Support: Typed class constants (Part 3)

- Fix the context sensitive lexer (Reserved keywords can be used as constant names)
- Get the constant names as the `PHP_STRING` token
- Add unit tests

Example:
```php
class Example {
    public const string|array array = [1, 2, 3], string = "example";
}
```

nb-php83-typed-class-constants-context-sensitive-lexer

### PHP 8.3 Support: Typed class constants (Part 4)

- Fix the mark occurences and the go to declaration features
- Add unit tests

### PHP 8.3 Support: Typed class constants (Part 5)

- Fix the Code Completion feature
- Add templates for constant statements
- Add unit tests

### PHP 8.3 Support: Typed class constants (Part 6)

- Fix the formatter (don't add spaces within parens of DNF types)
- Fix the `UnusableTypeHintError`
- Add unit tests

nb-php83-typed-class-constants-unusable-type-hint-error

### CODE REVIEW

File: php/php.editor/test/unit/data/testfiles/completion/lib/php74/testTypedProperties20Class/typedProperties20Class.php.testTypedProperties20Class_08.completion [hunk 1]
```diff
@@ -6,26 +6,17 @@ PACKAGE    Foo                             [PUBLIC]   null
 CLASS      MyClass                         [PUBLIC]   Bar
 CLASS      TypedPropertiesClass            [PUBLIC]   Foo
 ------------------------------------
-KEYWORD    abstract                                   null
```
Comment (junichi11) [inline php/php.editor/test/unit/data/testfiles/completion/lib/php74/testTypedProperties20Class/typedProperties20Class.php.testTypedProperties20Class_08.completion:9]: These are unnecessary in this caret position.

File: php/php.editor/test/unit/data/testfiles/completion/lib/php74/testTypedProperties20Class/typedProperties20Class.php.testTypedProperties20Class_18.completion [hunk 2]
```diff
@@ -1,14 +1,22 @@
 Code completion result for source line:
 public const |CONSTANT = "constant";
 (QueryType=COMPLETION, prefixSearch=true, caseSensitive=true)
+PACKAGE    Bar                             [PUBLIC]   null
```
Comment (junichi11) [inline php/php.editor/test/unit/data/testfiles/completion/lib/php74/testTypedProperties20Class/typedProperties20Class.php.testTypedProperties20Class_18.completion:4]: A constant declaration may have types.

File: php/php.editor/test/unit/data/testfiles/formatting/php82/dnfTypes_01.php.formatted [hunk 3]
```diff
@@ -69,7 +69,7 @@ class ClassY {
     private (ClassX&ClassZ)|(ClassY&ClassZ)|ClassX $privateYField;
     protected ClassX|(ClassY&ClassZ)|ClassY $protectedYField;
     public static (ClassY&ClassZ)|ClassX $publicStaticYField;
-    private static ( ClassY&ClassZ)|ClassY $privateStaticYField;
+    private static (ClassY&ClassZ)|ClassY $privateStaticYField;
```
Comment (junichi11) [inline php/php.editor/test/unit/data/testfiles/formatting/php82/dnfTypes_01.php.formatted:72]: This was a bug.

**Comment by junichi11** [comment]
@tmysik Could you please have a look at each commit when you have time?

File: php/php.editor/src/org/netbeans/modules/php/editor/model/impl/ClassConstantElementImpl.java [hunk 4]
```diff
@@ -29,9 +30,10 @@
 import org.netbeans.modules.php.editor.model.nodes.ClassConstantDeclarationInfo;
 
 class ClassConstantElementImpl extends ModelElementImpl implements ClassConstantElement {
-    private String typeName;
-    private final String value;
 
+    private final String typeName;
+    private final String value;
+    private final String declaedType;
```
Comment (tmysik) [inline php/php.editor/src/org/netbeans/modules/php/editor/model/impl/ClassConstantElementImpl.java:36]: Typo: `decla[r]edType`

File: php/php.editor/test/unit/data/testfiles/formatting/php82/dnfTypes_01.php.formatted (hunk 3 above)
Comment (tmysik) [reply php/php.editor/test/unit/data/testfiles/formatting/php82/dnfTypes_01.php.formatted:72]: I see.

**Review by tmysik — APPROVED** [review]
Looks good to me, thanks again for all the tests, as always :+1:

File: php/php.editor/src/org/netbeans/modules/php/editor/model/impl/ClassConstantElementImpl.java (hunk 4 above)
Comment (junichi11) [reply php/php.editor/src/org/netbeans/modules/php/editor/model/impl/ClassConstantElementImpl.java:36]: Oops. I'll fix it later. Thanks!

Comment (junichi11) [reply php/php.editor/src/org/netbeans/modules/php/editor/model/impl/ClassConstantElementImpl.java]: Fixed.

**Pull request merged**